api/rosetta/data_block_integration_test.go

```diff
@@ -0,0 +1,563 @@
+
+	t.Helper()
+
+	return func(tx *object.Transaction) {
+
+		assert.Equal(t, tx.ID.Hash, hash)
+		assert.Equal(t, len(tx.Operations), 2)
+
+		relatedOperations := make(map[uint]uint)
+
+		for _, op := range tx.Operations {
+
+			// save related operation IDs in a map so we can cross reference them
+			// TODO: check - is this special-casing the test too much perhaps?
+			// operation can have multiple related operations,
+			// but also this function does state that it's verifying a single transfer
+
+			if assert.Len(t, op.RelatedIDs, 1) {
+				relatedOperations[op.ID.Index] = op.RelatedIDs[0].Index
+			}
+
+			// validate operation and status
+			assert.Equal(t, op.Type, dps.OperationTransfer)
+			assert.Equal(t, op.Status, dps.StatusCompleted)
+
+			// validate currency
+			assert.Equal(t, op.Amount.Currency.Symbol, dps.FlowSymbol)
+			assert.Equal(t, op.Amount.Currency.Decimals, uint(dps.FlowDecimals))
+
+			// validate address
+			address := op.AccountID.Address
+			if address != from && address != to {
+				t.Errorf("unexpected account address (%v)", address)
+			}
+
+			// validate transfered amount
+			wantValue := strconv.FormatInt(amount, 10)
+			if address == from {
+				wantValue = "-" + wantValue
+			}
+
+			assert.Equal(t, op.Amount.Value, wantValue)
+		}
+
+		// cross-reference related operations - verify that the related operation backlinks to the original one
+		for id, relatedID := range relatedOperations {
+			assert.Contains(t, relatedOperations, relatedID)
+			assert.Equal(t, id, relatedOperations[relatedID])
+		}
+	}
+}
+
+func validateBlockID(t *testing.T, height uint64, hash string) blockIDValidationFn {
+
+	t.Helper()
+
+	return func(blockID identifier.Block) {
+		assert.Equal(t, height, blockID.Index)
+		assert.Equal(t, hash, blockID.Hash)
+	}
+}
+
+// blockID() looks like a natural function name, but don't want to occupy the variable name either
+func knownBlockID(height uint64) string {
+
+	// NOTE: map would be cleaner, but would be created on each call which seems wasteful
+
+	switch height {
+
+	case 0:
+		return "d47b1bf7f37e192cf83d2bee3f6332b0d9b15c0aa7660d1e5322ea964667b333"
+	case 1:
+		return "9eac11ab78ebb9650803eea70a48399f772c64892823a051298d445459cdbc46"
+	case 12:
+		return "9035c558379b208eba11130c928537fe50ad93cdee314980fccb695aa31df7fc"
+	case 13:
+		return "af528bb047d6cd1400a326bb127d689607a096f5ccd81d8903dfebbac26afb23"
+	case 42:
+		return "91c00b22dc9b84281d293f6e1ff680133239addd8b0220a244554e1d96aed8e0"
+	case 43:
+		return "dab186b45199c0c26060ea09288b2f16032da40fc54c81bb2a8267a5c13906e6"
+	case 44:
+		return "810c9d25535107ba8729b1f26af2552e63d7b38b1e4cb8c848498faea1354cbd"
+	case 424:
+		return "6af26621eca92babda2df3ebcd2fe269946b3bf208183569258630e64486831d"
+	case 425:
+		return "594d59b2e61bb18b149ffaac2b27b0efe1854f6795cd3bb96a443c3676d78683"
+
+	default:
+		return ""
+	}
+
+}
```

Comment:
It would make a lot more sense to me to return the full headers here, where you have everything that belongs to a block together. Then you can just use `knownBlock(height).ID()` for example, or `blockRequest(header)`.

Reply:
Should these headers be initialized by hand?

e.g.
```go
return flow.Header{
  ChainID: dps.FlowTestnet,
  ParentID: "xyz",
  Height: 45, 
  // rest of stuff here
}
```

I know I could get them programmatically from the index snapshot we load, but I dislike the notion of relying on another piece of code to be an accurate 'source of truth' for the tests so I wanted to check..

Reply:
https://github.com/optakt/flow-dps/pull/163/files#r651685459